### ISSUE TITLE
[BUG] Error in adaptive_random_forests.py: class votes carry a NaN value

### ISSUE THREAD

**Issue by Aashi-Jindal** [scikit-multiflow/scikit-multiflow#165]
"adaptive_random_forests.py" fails for kdd dataset. For other datasets, it works fine. The following error shows:

/scratch/root/anaconda/anaconda3/lib/python3.7/site-packages/skmultiflow/bayes/utils.py:31: RuntimeWarning: invalid value encountered in double_scalars
  votes[class_index] = observed_class_val / observed_class_sum
**{0: nan}** is the value of variable votes for an instance when this fails.
Traceback (most recent call last):
  File "adaptive_random_forests.py", line 67, in <module>
    pred_ = clf.predict(x)[0]
  File "/scratch/root/anaconda/anaconda3/lib/python3.7/site-packages/skmultiflow/meta/adaptive_random_forests.py", line 268, in predict
    y_proba = self.predict_proba(X)
  File "/scratch/root/anaconda/anaconda3/lib/python3.7/site-packages/skmultiflow/meta/adaptive_random_forests.py", line 300, in predict_proba
    y_proba_current = self.ensemble[i].predict_proba(X)
  File "/scratch/root/anaconda/anaconda3/lib/python3.7/site-packages/skmultiflow/meta/adaptive_random_forests.py", line 514, in predict_proba
    return self.classifier.predict_proba(X)
  File "/scratch/root/anaconda/anaconda3/lib/python3.7/site-packages/skmultiflow/trees/hoeffding_tree.py", line 445, in predict_proba
    votes = normalize_values_in_dict(votes, inplace=False)
  File "/scratch/root/anaconda/anaconda3/lib/python3.7/site-packages/skmultiflow/utils/utils.py", line 58, in normalize_values_in_dict
    raise ValueError('Can not normalize, normalization factor is NaN')
ValueError: Can not normalize, normalization factor is NaN

The snippet of code is as follows:

```python

    import numpy as np
    from skmultiflow.meta import AdaptiveRandomForest
    
        clf = AdaptiveRandomForest(n_estimators=n_estimators, random_state=1)
        clf.partial_fit(data[0].reshape(1,data.shape[1]), [label[0]], classes=np.unique(label))

        preds = []

        for i in range(data.shape[0]):
            if (i+1)%int((len(label)/20)) == 0:
                print('%d%%' % int(np.round((i+1.)/len(label)*100, 0)))
            x = data[i].reshape(1, data.shape[1])
            y = np.array([int(label[i])])
            pred_ = clf.predict(x)[0]
            preds.append(pred_)
            clf.partial_fit(x, y)
```

The kdd dataset has been downloaded from https://github.com/vlosing/driftDatasets, where labels have been mapped to integer values. The map of labels is as follows:
normal:0
buffer_overflow:1
loadmodule:2
perl:3
neptune:4
smurf:5
guess_passwd:6
pod:7
teardrop:8
portsweep:9
ipsweep:10
land:11
ftp_write:12
back:13
imap:14
satan:15
phf:16
nmap:17
multihop:18
warezmaster:19
warezclient:20
spy:21
rootkit:22

**Comment by smastelini** [comment]
Hello @Aashi-Jindal, thanks for pointing out that. Could you provide a complete MWE for this problem?

**Comment by Aashi-Jindal** [comment]
Please find below MWE:
```python
import numpy as np
import logging
import sys
import os
from skmultiflow.meta import AdaptiveRandomForest


if __name__ == '__main__':

    logging.basicConfig(format='%(message)s', level=logging.INFO)
    #Read data and labels
    dataN = sys.argv[2]
    dataPath = os.path.join('driftDatasets-master', sys.argv[1], dataN)
    datafile = os.path.join(dataPath, sys.argv[3])
    logging.info(datafile)
    labelfile = os.path.join(dataPath, sys.argv[4])

    data = np.genfromtxt(datafile, delimiter=sys.argv[5])
    logging.info(f'{data.shape[0]} samples')
    logging.info(f'{data.shape[1]} dimensions')
    label = np.genfromtxt(labelfile, dtype=np.int)
    ######################################################################

    #choose no. of estimators
    n_estimators = 10

    clf = AdaptiveRandomForest(n_estimators=n_estimators, random_state=1)
    clf.partial_fit(data[0].reshape(1,data.shape[1]), [label[0]], classes=np.unique(label))

    logging.info('applying model on dataset')
    for i in range(data.shape[0]):
        if (i+1)%int((len(label)/20)) == 0:
            logging.info('%d%%' % int(np.round((i+1.)/len(label)*100, 0)))
        x = data[i].reshape(1, data.shape[1])
        y = np.array([int(label[i])])
        pred_ = clf.predict(x)[0]
        clf.partial_fit(x, y)

```

Also, the issue appears after 5% of the dataset has been processed for streaming.
https://drive.google.com/drive/folders/1J8izTUEfEpl6cZHkjXTvIDreFC61dr2h?usp=sharing contains the kdd dataset along with labels used for this.

**Comment by smastelini** [comment]
Hello @Aashi-Jindal,

To update you on the bug tracking. I have been able to run your code and reproduce the problem.
Nevertheless, I haven't found the root of this strange behavior yet. I'll continue performing some tests to identify the source.

@jacobmontiel, do you have any hypothesis for this case?

**Comment by jacobmontiel** [comment]
Looking into the error log, this seems actually related to the Hoeffding Tree and not to the Adaptive Random Forest:

> File "/scratch/root/anaconda/anaconda3/lib/python3.7/site-packages/skmultiflow/trees/hoeffding_tree.py", line 445, in predict_proba
votes = normalize_values_in_dict(votes, inplace=False)

The deepest place where the bug could be triggered is:

`LearningNodeNBAdaptive.get_class_votes` -> `utils.def do_naive_bayes_prediction` -> `GaussianNumericAttributeClassObserver.probability_of_attribute_value_given_class` -> `GaussianEstimator.probability_density`

The tricky part is to catch where/when/how the `votes` are set to `NaN`...

It might be useful for debugging to use the `scipy` method instead:

```python
from scipy.stats import norm
norm.pdf(x, mean, std_dev)
```

**Note:** We are currently not using this method since it is slower in our setting.

**Comment by smastelini** [comment]
Following the suggestion of @jacobmontiel, I ran the example setting `leaf_predict` to `mc` (majority class) in ARF. The code processed all the data without errors.

Hence, the current hypothesis is that the problem lies in the Naive Bayes code. Further tests are needed to isolate the problem.

**Comment by smastelini** [comment]
Closing since it was addressed in #195.

**Comment by iyuyong** [comment]
A:\Program Files\Anaconda3\lib\site-packages\skmultiflow\trees\gaussian_estimator.py:36: RuntimeWarning: overflow encountered in ubyte_scalars
  self._mean += weight * (value - last_mean) / self._weight_sum
A:\Program Files\Anaconda3\lib\site-packages\skmultiflow\trees\gaussian_estimator.py:37: RuntimeWarning: overflow encountered in ubyte_scalars
  self._variance_sum += weight * (value - last_mean) * (value - self._mean)
